### PR TITLE
support env var expansion in envoy statsd urls

### LINKED ISSUE #8561: Support dynamic addresses for envoy proxy statsd_url configurations
#### Feature Description

Either/or:
- Provide the ability to resolve the statsd_url host address from an existing environment variable, e.g.

```
envoy_statsd_url: udp://${HOST_IP}:1234
```

- Provide the ability to specify custom environment variables for the consul-connect-inject-init sidecar
This would provide a way to make use of the `$ENV_VAR_NAME` configuration option specified here: (https://www.consul.io/docs/connect/proxies/envoy#envoy_statsd_url)

#### Use Case(s)

`envoy_statsd_url` and `envoy_dogstatsd_url` must take an ip address and not a hostname. This presents problems when using daemonset statsd agents, because the config for each proxy must point to the local `HOST_IP`. The Connect Inject workflow does not provide any method to dynamically inject the appropriate address into the statsd_url configurations.

While the `consul connect enoy` command supports reading the entire config value from an environment variable,  there's no meaningful way to populate such a variable in the injection sidecar container. The environment variable `HOST_IP` is already present in the container, but the dynamic env var configuration option requires that the target env var contain the entire configuration url, e.g. `udp://ip:port`. 

The code [here](https://github.com/hashicorp/consul/blob/master/command/connect/envoy/bootstrap_config.go#L267) could be updated to support expanding environment variables in the config string. I believe this option would be the easiest to implement and I'd be happy to submit a PR with the changes. My only concern is if there was a reason this functionality was explicitly considered yet not implemented initially. From the docs: "It is not currently possible to use environment variables as only part of the URL." If there's a security concern or something, perhaps maintaining an allowlist of accepted env vars would be sufficient, e.g. only allow expansion of `HOST_IP`. 

Alternatively, providing the ability to pass additional arbitrary env vars into the inject container would allow users to leverage the intended dynamic environment config option for these settings. This seems pretty messy and involved.

### PR DESCRIPTION
**Pull request by woz5999**
Fixes #8561

### CODE REVIEW

File: command/connect/envoy/bootstrap_config.go [hunk 1]
```diff
@@ -305,6 +307,18 @@ func (c *BootstrapConfig) generateStatsSinkJSON(name string, addr string) (strin
...
 }
 
+func (c *BootstrapConfig) statsSinkEnvMapping(s string) string {
+	allowedStatsSinkEnvVars := map[string]bool{
+		"HOST_IP": true,
+	}
+
+	if !allowedStatsSinkEnvVars[s] {
+		// if the specified env var isn't explicitly allowed, unexpand it
+		return fmt.Sprintf("${%s}", s)
```
Comment (woz5999) [inline command/connect/envoy/bootstrap_config.go:317]: Could also just return an empty string. no strong preference here; either way, if someone enters a disallowed env var, it's going to fail `url.Parse`, so the result here is just cosmetic.

File: command/connect/envoy/bootstrap_config.go [hunk 2]
```diff
@@ -305,6 +307,18 @@ func (c *BootstrapConfig) generateStatsSinkJSON(name string, addr string) (strin
 	}`, nil
 }
 
+func (c *BootstrapConfig) statsSinkEnvMapping(s string) string {
+	allowedStatsSinkEnvVars := map[string]bool{
```
Comment (woz5999) [inline command/connect/envoy/bootstrap_config.go:311]: this seemed like the simplest place to put this allowlist. happy to move elsewhere if desired. 
the original implementation explicitly didn't support expanding arbitrary env vars, so the purpose here to minimize the potential side effects while still providing the ability to dynamically set this. happy to remove this functionality and just use a plain `os.ExpandEnv` if that's preferable.

**Comment by woz5999** [comment]
@rboyer @mkeeler what's the best way to get this change reviewed?

**Comment by woz5999** [comment]
@mkeeler @i0rek what's the best way to get this change reviewed?

**Comment by woz5999** [comment]
@dnephin @s-christoff what's the best way to get this change reviewed?

**Comment by woz5999** [comment]
@mkcp what's the best way to get this change reviewed?

**Review by dnephin — COMMENTED** [review]
Thank you for the PR! Sorry it took us a while to review it.

I think this looks good! I rebased the PR to resolve some conflicts, updated the test expected value because it had changed due to other changes, and also added a changelog entry.

If the tests still pass I think we are good to merge.

**Review by dnephin — APPROVED** [review]

**Pull request merged**